lib/src/space/mod.rs

```diff
@@ -128,6 +128,12 @@ pub trait Space {
             .collect()
     }
 
+    /// Returns the number of Atoms in the space, or None if this can't be determined
+    fn atom_count(&self) -> Option<usize>;
```

Comment:
My main point that `atom_count` seems to be duplicating part of the iterator API. In Rust we could eliminate duplication by implementing `size_hints()` for the `SpaceIter`. Thus when there is a way to calculate the number of atoms then space implementer could do it in `size_hints`. In C API it looks like we anyway need an additional function for doing this. What do you think?

Reply:
My concern with using `size_hint()` is that it isn't authoritative.  For example `Vec` still has a `len()` method because the caller is told not to rely on `size_hint()` for anything other than optimization (buffer pre-allocation).  So I believe they serve different purposes, and we should keep the `atom_count()` method.

Reply:
I think the ideas behind this sub-thread just clicked with me.  I just realized you may have been referring to `Iterator.count()`.  It's a poor substitute for a length method because it needs to actually iterate all the items (order n) rather than just reading a value (constant time).

But then I got to thinking about how the current GroundingSpace implementation uses `count()` to implement `atom_count` and how that's no better.  At the time I was thinking that a future space could track its atoms more directly to make this a direct access too.  However, I also realized the current implementation of GroundingSpace is iterable, but also doesn't have an atom_count ready to go.  So in that case `atom_count()` might return None, but `atom_iter()` might return a valid iterator.  So relying on atom_count to determine whether atom_iterator is valid (as the C API does) is a bad design.

Iterable spaces with no definitive `atom_count` might make sense for some conceptual spaces, but that means I need to tweak the C API to return an error irrespective of the result of `atom_count`

Reply:
Maybe it is more general than adding or not adding `atom_count`. New implementations of `Space` will not be like `Vec`, they will be more like a database. In that sense probably both `atom_count()` and `atom_iter()` are not good candidates for the `Space` API. And both of them are not critical for MeTTa, but sometimes are useful for debugging. 

We could provide specific APIs via downcasting. But both Rust and C don't have a built-in downcasting mechanism, so it should be implemented manually. Methods which return `Option<T>` are a kind of downcasting. We could invent a better way to downcast or keep current implementation.

If we look at this issue from such perspective then `Space` should have default implementations for `atom_count` and `atom_iter`. The default implementation should notify the user that method is not supported. I see two options here: use `Option<T>` or `Result<T, E>`. Returning error seems to be more traditional to me, but actually I don't have a strong opinion here. Taking into account it is simpler to use optional in C API may be we should just use `Option<T>` everywhere.

Reply:
I committed 70e3b1b to decouple the implementations of space_iterate from space_atom_count, at the C and Python levels.

I didn't see your comment about the default implementation until just now, but it's a good idea and I'll add it right now.

Reply:
Done in 7961af2

Reply:
> ... Option<T> or Result<T, E>. Returning error seems to be more traditional to me, but actually I don't have a strong opinion here. Taking into account it is simpler to use optional in C API may be we should just use Option<T> everywhere.

My personal philosophy on Result vs. Option is that Result is for exceptional / unexpected situations while Option is for normal situations where a None value is possible.  Like for example from the `next()` method on an iterator when the last item has been encountered.  A Result lets the caller pass the exception up (using the '?' operator) but the Option must be explicitly handled by the caller at the call site.

In this case, if the caller is going to use an... optional... feature of the Space interface, they must handle the cases where that API is not implemented.  So I think an Option is appropriate.

Reply:
Thanks Luke, I also didn't see you previous comment because wrote mine in parallel. Yes, `atom_count` and `atom_iter` are not always can be implemented effectively at the same time. I like your solution in https://github.com/trueagi-io/hyperon-experimental/commit/70e3b1ba8a87c93f1fd1d7e9b7675cceaef06bbc